### ISSUE TITLE
Resolving 'Unknown Compiler' Error in clangd Plugin: Troubleshooting and Solutions

### ISSUE THREAD

**Issue by DYAAL** [clangd/vscode-clangd#624]
The clangd plugin is reporting an error: "In included file: Unknown compiler." However, my compilation and other functionalities work fine. How can I get rid of this error?

![image](https://github.com/clangd/vscode-clangd/assets/126035353/56589d31-4d51-4fcc-b606-2be004a5ef00)

**Comment by HighCommander4** [comment]
Please share [clangd logs](https://clangd.llvm.org/troubleshooting.html#gathering-logs)

**Comment by DYAAL** [comment]
> 请分享[clangd 日志](https://clangd.llvm.org/troubleshooting.html#gathering-logs)

![image](https://github.com/clangd/vscode-clangd/assets/126035353/6692fc47-f635-46ee-bbf6-0ef926427e6b)
The clangd log is shown as above.

This question might be related to the fact that I'm using cross-compilation. Here's my clangd configuration:
![image](https://github.com/clangd/vscode-clangd/assets/126035353/912e6f6e-0945-42f2-9690-87481fa42e31)

**Comment by HighCommander4** [comment]
> The clangd log is shown as above.

That's not a log, that's a screenshot of something (the hover shown on the diagnostic I guess).

Please take another look at the instructions I linked to for getting logs. They are found in the "Output" view. Please share the full log in text form.

> Here's my clangd configuration:

I think when you provide multiple `--query-driver` arguments, a later one overrides the earlier one.

If you want to specify multiple patterns, you can do that by separating them with a comma, e.g. 

```
--query-driver=/path/one,/path/two
```

**Comment by DYAAL** [comment]
> > The clangd log is shown as above.
> 
> That's not a log, that's a screenshot of something (the hover shown on the diagnostic I guess).
> 
> Please take another look at the instructions I linked to for getting logs. They are found in the "Output" view. Please share the full log in text form.
> 
> > Here's my clangd configuration:
> 
> I think when you provide multiple `--query-driver` arguments, a later one overrides the earlier one.
> 
> If you want to specify multiple patterns, you can do that by separating them with a comma, e.g.
> 
> ```
> --query-driver=/path/one,/path/two
> ```
here is my [clangd logs](https://clangd.llvm.org/troubleshooting.html#gathering-logs), I hope this may help you to find the problem.
[clangd_log.txt](https://github.com/clangd/vscode-clangd/files/15092941/clangd_log.txt)

Thanks a lot for your help!

**Comment by HighCommander4** [comment]
The "unknown compiler" error is coming from an `#error` directive in the file `d:/001_my_Rapier/my_Rapier/Drivers/CMSIS/Include/cmsis_compiler.h`. Could you share the contents of this file?

**Comment by DYAAL** [comment]
> cmsis_compiler.h

Certainly. Additionally, the clangd plugin will report an error "Call to undeclared function 'XXXXX'; ISO C99 and later do not support implicit function declarations", but my code compiles without issues. This is also somewhat perplexing.
![image](https://github.com/clangd/vscode-clangd/assets/126035353/c9c5b6af-ad81-40ac-a166-72a072175170)
GitHub doesn't allow directly uploading .h files, so I renamed it as a .txt file before uploading.







[cmsis_compiler.txt](https://github.com/clangd/vscode-clangd/files/15104561/cmsis_compiler.txt)

**Comment by HighCommander4** [comment]
> [cmsis_compiler.txt](https://github.com/clangd/vscode-clangd/files/15104561/cmsis_compiler.txt)

It looks like this file expects that you're using one of a specific set of compilers, and clang doesn't seem to be one of them. So, it's possible that this project does not support clang as a compiler, which might mean not supporting clang-based tools like clangd either.

That said, I notice that one of the preprocessor branches checks for `__GNUC__` (which is defined by the GCC compiler), and I believe clang does define this macro in an attempt to be compatible with GCC.

What compiler are you actually building your code with? Are you able to build it with clang?

**Comment by DYAAL** [comment]
> It looks like this file expects that you're using one of a specific set of compilers, and clang doesn't seem to be one of them. So, it's possible that this project does not support clang as a compiler, which might mean not supporting clang-based tools like clangd either.
> 
> That said, I notice that one of the preprocessor branches checks for `__GNUC__` (which is defined by the GCC compiler), and I believe clang does define this macro in an attempt to be compatible with GCC.
> 
> What compiler are you actually building your code with? Are you able to build it with clang?

This project is an ARM project and is compiled through cross-compilation. However, the compiler used is probably the one included with Keil, named armcc.exe, and this might have caused some issues. By the way, can clang support cross-compilation? I haven't tried this feature before, but it's always a hassle to build files like Makefile or Cmake from scratch.
